### PR TITLE
Network refactor

### PR DESCRIPTION
**Pull request by TheCharlatan**
Based on #73 

Refactors the Socket and Blockbook functions to be more like the networking code of the old, battle-tested edge-currency-bitcoin plugin.
 
This commit is meant as pre-cursor work to lay the foundation for a task queue and multiple sockets.

The new networking code is designed to use fewer callbacks and more promises and events, favors maps over objects with dynamic keys, has a more robust keep alive / heartbeat system that takes existing queries into account, has a cleaner subscription interface and allows for the potential usage of a task queue and multiple sockets.

Further, some more error handling is introduced, though this will need further improvement.

Some more test cases in the block book tests are added to test these new capabilities. Note that this work does not change the interface of the blockbook functions though.

### CODE REVIEW

File: src/common/utxobased/network/BlockBook.ts [hunk 1]
```diff
@@ -307,28 +241,28 @@ export function makeBlockBook(config: BlockBookConfig): BlockBook {
     })
   }
 
-  function watchBlocks(cb: Callback): void {
-    blockWatcherCallback = cb
+  async function watchBlocks(cb: Callback): Promise<void> {
+    const socketCb = async (value: INewBlockResponse): Promise<void> => {
+      // eslint-disable-next-line no-void
+      await cb()
```
Comment (TheCharlatan) [inline src/common/utxobased/network/BlockBook.ts:247]: Follow-up: Remove callback, just use even emitter.

File: src/common/utxobased/network/BlockBook.ts [hunk 2]
```diff
@@ -307,28 +241,28 @@ export function makeBlockBook(config: BlockBookConfig): BlockBook {
...
+      method: 'subscribeNewBlock',
+      params: {},
+      cb: socketCb
+    })
   }
 
   function watchAddresses(
     addresses: string[],
-    cb?: (response: INewTransactionResponse) => void
+    cb: (response: INewTransactionResponse) => Promise<void>
```
Comment (TheCharlatan) [inline src/common/utxobased/network/BlockBook.ts:259]: Follow-up: Remove callback, just use event emitter.

File: src/common/utxobased/network/Socket.ts [hunk 3]
```diff
@@ -1,169 +1,316 @@
...
 }
 
-export interface InnerSocketCallbacks extends SocketCallbacks {
-  onOpen: () => void
-}
-
-export function makeSocket(uri: string, config?: SocketConfig): Socket {
+export function makeSocket(uri: string, config: SocketConfig): Socket {
   let socket: InnerSocket | null
+  const { emitter, onQueueSpace, queueSize = 5, walletId = '' } = config
```
Comment (TheCharlatan) [inline src/common/utxobased/network/Socket.ts:54]: Follow-up: Remove the walletId and if needed use in ping/healthCheck callback.

File: src/common/utxobased/network/Socket.ts [hunk 4]
```diff
@@ -1,169 +1,316 @@
...
 }
 
 interface SocketConfig {
-  callbacks?: SocketCallbacks
+  queueSize?: number
+  timeout?: number
+  walletId?: string
+  emitter: Emitter
+  onQueueSpace: () => potentialWsTask
+  healthCheck: () => Promise<object> // function for heartbeat, should submit task itself
```
Comment (TheCharlatan) [inline src/common/utxobased/network/Socket.ts:44]: Add edgelog to constructor.

File: src/common/utxobased/network/Socket.ts [hunk 5]
```diff
@@ -1,169 +1,316 @@
...
-}
 
-function setupWS(uri: string, callbacks?: InnerSocketCallbacks): InnerSocket {
-  const ws = new WS(uri)
+  const onTimer = (): void => {
+    const now = Date.now() - TIMER_SLACK
+    if (lastKeepAlive + KEEP_ALIVE_MS < now) {
+      lastKeepAlive = now
+      // eslint-disable-next-line no-void
+      void config
```
Comment (TheCharlatan) [inline src/common/utxobased/network/Socket.ts:189]: Check again, probably does not exception, swap .catch and .then order below.

File: src/common/utxobased/network/Socket.ts [hunk 6]
```diff
@@ -1,169 +1,316 @@
...
+        if (message == null) {
+          throw new Error(`Bad response id in ${messageJson}`)
+        }
+        pendingMessages.delete(id)
+        const { error } = json
+        try {
+          if (error != null) {
+            const errorMessage =
+              error.message != null
+                ? error.message.split('\n')[0]
```
Comment (TheCharlatan) [inline src/common/utxobased/network/Socket.ts:251]: Remove newline and split.

File: src/common/utxobased/network/nodejsWS.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+  uri: string,
+  callbacks: InnerSocketCallbacks
+): InnerSocket {
+  const ws = new WS(uri)
+
+  ws.on('open', () => {
+    callbacks.onOpen()
+  })
+  ws.on('message', (data: WS.Data) => {
+    // eslint-disable-next-line @typescript-eslint/no-base-to-string
```
Comment (TheCharlatan) [inline src/common/utxobased/network/nodejsWS.ts:15]: Use JSON.stringify

File: src/common/utxobased/network/socketQueue.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
+const QUEUE_JOBS_PER_RUN = 3
```
Comment (TheCharlatan) [inline src/common/utxobased/network/socketQueue.ts:1]: rename file to taskQueue.ts

**Comment by TheCharlatan** [comment]
Broadcast Tx should eventually return something, not just be awaited.

**Review by thehobbit85 — APPROVED** [review]
